File: Remote/Source/Networking/ITachDevice.m
//
//  NDiTachDevice.m
//  Remote
//
//  Created by Jason Cardwell on 9/4/14.
//  Copyright (c) 2014 Moondeer Studios. All rights reserved.
//

#import "ITachDevice.h"
#import "ITachDeviceViewController.h"

static int ddLogLevel   = LOG_LEVEL_DEBUG;
static int msLogContext = LOG_CONTEXT_CONSOLE;
#pragma unused(ddLogLevel,msLogContext)

@interface ITachDevice ()

@property (nonatomic, copy, readwrite) NSString * make;
@property (nonatomic, copy, readwrite) NSString * model;
@property (nonatomic, copy, readwrite) NSString * status;
@property (nonatomic, copy, readwrite) NSString * configURL;
@property (nonatomic, copy, readwrite) NSString * revision;
@property (nonatomic, copy, readwrite) NSString * pcbPN;
@property (nonatomic, copy, readwrite) NSString * pkgLevel;
@property (nonatomic, copy, readwrite) NSString * sdkClass;

@end

@interface ITachDevice (CoreDataGenerated)

@property (nonatomic) NSString * primitiveMake;
@property (nonatomic) NSString * primitiveModel;
@property (nonatomic) NSString * primitiveStatus;
@property (nonatomic) NSString * primitiveConfigURL;
@property (nonatomic) NSString * primitiveRevision;
@property (nonatomic) NSString * primitivePcbPN;
@property (nonatomic) NSString * primitivePkgLevel;
@property (nonatomic) NSString * primitiveSdkClass;

@end


@implementation ITachDevice

@dynamic pcbPN, pkgLevel, sdkClass, make, model, status, configURL, revision;

/// detailViewController
/// @return ITachDeviceViewController *
- (ITachDeviceViewController *)detailViewController {
  return [ITachDeviceViewController controllerWithItem:self];
}

/// editingViewController
/// @return ITachDeviceViewController *
- (ITachDeviceViewController *)editingViewController {
  return [ITachDeviceViewController controllerWithItem:self editing:YES];
}


/// setModel:
/// @param model
- (void)setModel:(NSString *)model {
  [self willChangeValueForKey:@"model"];
  self.primitiveModel = model;
  [self didChangeValueForKey:@"model"];
  if (self.isNameAutoGenerated) {
    NSMutableString * name = [@"" mutableCopy];
    if (StringIsNotEmpty(self.make)) [name appendString:self.make];
    if (StringIsNotEmpty(model)) {
      if (StringIsNotEmpty(name)) [name appendString:@"-"];
      [name appendString:model];
    }
    if (StringIsNotEmpty(name)) [self setPrimitiveValue:name forKey:@"name"];
  }
}

/// setMake:
/// @param make
- (void)setMake:(NSString *)make {
  [self willChangeValueForKey:@"make"];
  self.primitiveMake = make;
  [self didChangeValueForKey:@"make"];
  if (self.isNameAutoGenerated) {
    NSMutableString * name = [@"" mutableCopy];
    if (StringIsNotEmpty(self.model)) [name appendString:self.model];
    if (StringIsNotEmpty(make)) {
      if (StringIsNotEmpty(name)) [name appendString:@"-"];
      [name appendString:make];
    }
    if (StringIsNotEmpty(name)) [self setPrimitiveValue:name forKey:@"name"];
  }
}

/// setConfigURL:
/// @param configURL
- (void)setConfigURL:(NSString *)configURL {

  [self willChangeValueForKey:@"configURL"];

  // Strip 'http://' if present
  self.primitiveConfigURL = ([configURL hasPrefix:@"http://"]
                             ? [configURL substringFromIndex:7]
                             : configURL);

  [self didChangeValueForKey:@"configURL"];


}

/// updateWithData:
/// @param data
- (void)updateWithData:(NSDictionary *)data {

  [super updateWithData:data];

  self.pcbPN      = data[@"pcb-pn"];
  self.pkgLevel   = data[@"pkg-level"];
  self.sdkClass   = data[@"sdk-class"];
  self.make       = data[@"make"];
  self.model      = data[@"model"];
  self.status     = data[@"status"];
  self.configURL  = data[@"config-url"];
  self.revision   = data[@"revision"];

}

/// JSONDictionary
/// @return MSDictionary *
- (MSDictionary *)JSONDictionary {

  MSDictionary * dictionary = [super JSONDictionary];

  dictionary[@"type"] = @"itach";

  SafeSetValueForKey(self.pcbPN,      @"pcb-pn",      dictionary);
  SafeSetValueForKey(self.pkgLevel,   @"pkg-level",   dictionary);
  SafeSetValueForKey(self.sdkClass,   @"sdk-class",   dictionary);
  SafeSetValueForKey(self.make,       @"make",        dictionary);
  SafeSetValueForKey(self.model,      @"model",       dictionary);
  SafeSetValueForKey(self.status,     @"status",      dictionary);
  SafeSetValueForKey(self.configURL,  @"config-url",   dictionary);
  SafeSetValueForKey(self.revision,   @"revision",    dictionary);

  return dictionary;

}

@end

File: DataModel/Model/ITachDevice.swift
//
//  ITachDevice.swift
//  Remote
//
//  Created by Jason Cardwell on 10/1/14.
//  Copyright (c) 2014 Moondeer Studios. All rights reserved.
//

import Foundation
import CoreData
import MoonKit

@objc(ITachDevice)
public class ITachDevice: NetworkDevice {

  public var configURL: String! {
    get {
      willAccessValueForKey("configURL")
      let url = primitiveValueForKey("configURL") as? String
      didAccessValueForKey("configURL")
      return url ?? ""
    }
    set {
      willChangeValueForKey("configURL")
      setPrimitiveValue(newValue.hasPrefix("http://") ? newValue[7..<newValue.length] : newValue, forKey: "configURL")
      didChangeValueForKey("configURL")
    }
  }

  private var generatedName: String? {
    var n = ""
      if model != nil && !model.isEmpty { n += model }
      if make != nil && !make.isEmpty {
        if !n.isEmpty { n += "-" }
        n += make
      }
      return n.isEmpty ? nil : n
  }

  public var make: String! {
    get {
      willAccessValueForKey("make")
      let m = primitiveValueForKey("make") as? String
      didAccessValueForKey("make")
      return m ?? ""
    }
    set {
      willChangeValueForKey("make")
      setPrimitiveValue(newValue, forKey: "make")
      didChangeValueForKey("make")
      if isNameAutoGenerated { if let n = generatedName { setPrimitiveValue(n, forKey: "name") } }
    }
  }

  public var model: String! {
    get {
      willAccessValueForKey("model")
      let m = primitiveValueForKey("model") as? String
      didAccessValueForKey("model")
      return m ?? ""
    }
    set {
      willChangeValueForKey("model")
      setPrimitiveValue(newValue, forKey: "model")
      didChangeValueForKey("model")
      if isNameAutoGenerated { if let n = generatedName { setPrimitiveValue(n, forKey: "name") } }
    }
  }

  @NSManaged public var pcbPN: String
  @NSManaged public var pkgLevel: String
  @NSManaged public var revision: String
  @NSManaged public var sdkClass: String
  @NSManaged public var status: String

  /**
  updateWithData:

  :param: data ObjectJSONValue
  */
  override public func updateWithData(data: ObjectJSONValue) {
    super.updateWithData(data)
    if let pcbPN     = String(data["pcbPN"]) { self.pcbPN = pcbPN }
    if let pkgLevel  = String(data["pkgLevel"]) { self.pkgLevel = pkgLevel }
    if let sdkClass  = String(data["sdkClass"]) { self.sdkClass = sdkClass }
    if let make      = String(data["make"]) { self.make = make }
    if let model     = String(data["model"]) { self.model = model }
    if let status    = String(data["status"]) { self.status = status }
    if let configURL = String(data["configURL"]) { self.configURL = configURL }
    if let revision  = String(data["revision"]) { self.revision = revision }
  }

  override public var jsonValue: JSONValue {
    var obj = ObjectJSONValue(super.jsonValue)!
    obj["type"] = "itach"
    obj["pcbPN"] = pcbPN.jsonValue
    obj["pkgLevel"] = pkgLevel.jsonValue
    obj["sdkClass"] = sdkClass.jsonValue
    obj["make"] = make.jsonValue
    obj["model"] = model.jsonValue
    obj["status"] = status.jsonValue
    obj["configURL"] = configURL.jsonValue
    obj["revision"] = revision.jsonValue
    return obj.jsonValue
  }

}
